[CORE-6807] `kafka`: change `offset_out_of_range` condition in `replicated_partition::prefix_truncate()`

WillemKauf opened the pull request:
Previously, we would return an `error_code::offset_out_of_range` for a request where `kafka_truncation_offset <= start_offset`.

This is a divergence from the current Kafka behaviour, and can lead to errors with certain Kafka clients/consumers.

Remove the condition that throws an `offset_out_of_range` error if `kafka_truncation_offset <= start_offset()`, instead returning `error_code::none`.

More context and details in the JIRA link below.

JIRA Link: https://redpandadata.atlassian.net/browse/CORE-6807

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v24.2.x
- [X] v24.1.x
- [ ] v23.3.x

## Release Notes

### Improvements

* Allows `DeleteRecords` requests from Kafka clients or `rpk topic trim-prefix` to be called with `truncation_offset <= start_offset` without returning an error. The request is instead treated as a no-op.

nvartolomei commented on src/v/kafka/server/replicated_partition.cc:
```diff
@@ -486,7 +486,7 @@ ss::future<error_code> replicated_partition::prefix_truncate(
   ss::lowres_clock::time_point deadline) {
     if (
-      kafka_truncation_offset <= start_offset()
+      kafka_truncation_offset < start_offset()
```
I'm thinking we should short-circuit and return success directly if `kafka_truncation_offset == start_offset()`. Why would we replicate a no-op command?

WillemKauf replied:
Added an early return for `kafka_truncation_offset == start_offset()`.

WillemKauf commented:
Force push to:
* Add early exit on `kafka_truncation_offset == start_offset()` in `replicated_partition::prefix_truncate()`
* Add `KafkaCliTools.delete_records()`
* Add `DeleteRecordsKafkaTest` to make use of `KafkaCliTools.delete_records()`, asserting we achieve parity with Kafka.

nvartolomei commented on tests/rptest/tests/delete_records_test.py:
```diff
@@ -775,3 +783,153 @@ class PartitionMoveExceptionReporter:
+        partition_watermark_lines = split_str[2:-1]
+        result_list: list[RpkTrimOffsetResponse] = []
+        for partition_watermark_line in partition_watermark_lines:
+            topic_partition_str, result_str = partition_watermark_line.strip(
```
extract the parsing into a function please otherwise this code is barely readable

nvartolomei replied:
maybe this parsing needs to go into KafkaCliTools?

WillemKauf replied:
Moved parsing back into `KafkaCliTools.delete_records()`.

nvartolomei commented on tests/rptest/tests/delete_records_test.py:
```diff
@@ -775,3 +783,153 @@ class PartitionMoveExceptionReporter:
         assert status.validator.out_of_scope_invalid_reads == 0
+
+
+class DeleteRecordsKafkaTest(RedpandaTest):
```
nit: DeleteRecordsKafkaCliToolsTest

WillemKauf replied:
Renamed test to your suggestion.

nvartolomei commented on tests/rptest/tests/delete_records_test.py:
```diff
@@ -775,3 +783,153 @@ class PartitionMoveExceptionReporter:
+        delete_records_json = self.make_delete_records_for_cli(
+            delete_records_tpos)
+
+        delete_records_result = self.execute_delete_records(
```
make/execute are always called together. just make it one (wrapper) method

nvartolomei replied:
this will make very easy to understand what the test is actually testing

make non-empty error_msg throw in that helper method or in execute_delete_records so you don't have to assert it is empty everywhere

where you need to assert that it throws, test it with `rptest.util.expect_exception`

WillemKauf replied:
Refactored `execute_kafka_delete_records()` to be the one wrapper method needed to make a request, and further cleaned up test.

nvartolomei commented on tests/rptest/tests/delete_records_test.py:
```diff
@@ -775,3 +783,153 @@ class PartitionMoveExceptionReporter:
+        else:
+            assert False, "unknown truncate point encountered"
+
+        do_test_body(truncate_offset, expect_error, expected_new_start_offset)
```
confused as to why this is extracted as an inner function?

WillemKauf replied:
Cleaned up test by removing inner function.

WillemKauf commented:
Force push to address Nicolae's comments:
* Renamed `DeleteRecordsKafkaTest` -> `DeleteRecordsKafkaCliToolsTest`
* Moved parsing of `DeleteRecords` result into `kafka_cli_tools.delete_records()`
* Refactored and cleaned up `test_delete_records_non_empty_topic()` further.

andrwng commented on src/v/kafka/server/replicated_partition.cc:
```diff
@@ -485,8 +485,12 @@ replicated_partition::get_leader_epoch_last_offset_unbounded(
 ss::future<error_code> replicated_partition::prefix_truncate(
   model::offset kafka_truncation_offset,
   ss::lowres_clock::time_point deadline) {
+    if (kafka_truncation_offset == start_offset()) {
```
`start_offset()` is not quite a trivial method. Consider caching it in a variable

andrwng commented on tests/rptest/tests/delete_records_test.py:
```diff
@@ -782,3 +784,128 @@ class PartitionMoveExceptionReporter:
+
+
+class DeleteRecordsKafkaCliToolsTest(RedpandaTest):
+    """
+    To assert DeleteRecords has the same semantics with a Kafka client.
+    Uses kafka-delete-records.sh and rpk trim-prefix directly to assert
+    we achieve parity with Kafka
+    """
```
nit: i'm not against doing this test still, but typically when we think of parity with Kafka we are thinking about server-side parity. In that regard, the _real_ test for parity is to spin up a kafka broker and use a single client to talk to both

WillemKauf commented:
Force push to address Andrew's comments:
* Added a mixin test `BaseDeleteRecordsTest` which utilizes both a `redpanda` cluster as well as a `KafkaService` to assert parity beyond just the CLIs.
* This test actually revealed that Kafka does not return an `offset_out_of_range` error for any `DeleteRecords` request with `truncation_offset <= start_offset`- amended the `redpanda` code to no-op for this entire range of values.

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/53078#01915dd0-b9b4-42ca-afa9-8e7a381f5281
<!--689428e3-ef7a-4261-b47e-9446c87d0799-->

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
/backport v24.1.x

dotnwat reviewed (commented):
Just guessing that we should have gotten a replication and maybe enterprise reviewer on this pr.

WillemKauf commented:
> Just guessing that we should have gotten a replication and maybe enterprise reviewer on this pr.

Ack, @bharathv, @oleiman, would you like to weigh in on this change?

oleiman reviewed (commented):
i think any divergence from kafka at the API layer is considered a bug from the enterprise team's perspective, so the spirit of the change seems legit. nice tests!

bharathv commented on src/v/kafka/server/replicated_partition.cc:
```diff
@@ -485,9 +485,11 @@ replicated_partition::get_leader_epoch_last_offset_unbounded(
-    if (
-      kafka_truncation_offset <= start_offset()
-      || kafka_truncation_offset > high_watermark()) {
+    if (kafka_truncation_offset <= start_offset()) {
```
nit: https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/cluster/Partition.scala#L1688

suggests that if truncation_offset < 0 it should throw a OffsetOutOfRangeException(), its an edge case so probably not a big deal.

WillemKauf replied:
Ah, there is even more nuance here than that: `-1` indicates a complete wipe up to `high_watermark`, while other values `< 0` are an `offset_out_of_range` error. Will have this covered in a follow up :+1:
